cpp/src/arrow/dataset/scanner.cc

```diff
@@ -428,6 +428,11 @@ class ARROW_DS_EXPORT AsyncScanner : public Scanner,
   Result<FragmentGenerator> GetFragments() const;
 
   std::shared_ptr<Dataset> dataset_;
+
+  // XXX if Scanner were truly single-use this would be a 1:1 relationship
```

Comment:
I'll write a follow up JIRA. Many of the unit tests assume a scanner can be reused, and it'll be a change in behavior in the Python API too since we expose Scanner there

Reply:
https://issues.apache.org/jira/browse/ARROW-13264